File: crypto/utils/file_helper.py
"""
File Helper Utilities for Hybrid Cryptography System
Handles metadata generation, folder management, and file operations
"""

import os
import json
import time
from datetime import datetime
from pathlib import Path

def create_project_folders(base_dir):
    """
    Create all necessary project folders
    
    Args:
        base_dir (str): Base directory for the project
        
    Returns:
        dict: Dictionary of folder paths
    """
    folders = {
        'files': os.path.join(base_dir, 'files'),
        'encrypted': os.path.join(base_dir, 'encrypted'), 
        'decrypted': os.path.join(base_dir, 'decrypted'),
        'metadata': os.path.join(base_dir, 'metadata'),
        'keys': os.path.join(base_dir, 'keys')
    }
    
    # Create all folders
    for folder_name, folder_path in folders.items():
        os.makedirs(folder_path, exist_ok=True)
        print(f"[INFO] Folder ready: {folder_name}/ -> {folder_path}")
    
    return folders

def generate_timestamp():
    """
    Generate timestamp string for file naming
    
    Returns:
        str: Timestamp in format YYYYMMDD_HHMMSS
    """
    return datetime.now().strftime("%Y%m%d_%H%M%S")

def generate_filename(original_name, suffix="", extension=".enc", include_timestamp=True):
    """
    Generate filename with optional timestamp and suffix
    
    Args:
        original_name (str): Original filename
        suffix (str): Optional suffix to add
        extension (str): File extension
        include_timestamp (bool): Whether to include timestamp
        
    Returns:
        str: Generated filename
    """
    base_name = Path(original_name).stem
    
    parts = [base_name]
    
    if suffix:
        parts.append(suffix)
        
    if include_timestamp:
        parts.append(generate_timestamp())
    
    filename = "_".join(parts) + extension
    return filename

def save_metadata(metadata_dict, metadata_file_path):
    """
    Save metadata to JSON file
    
    Args:
        metadata_dict (dict): Metadata dictionary
        metadata_file_path (str): Path to save metadata file
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        os.makedirs(os.path.dirname(metadata_file_path), exist_ok=True)
        
        with open(metadata_file_path, 'w', encoding='utf-8') as f:
            json.dump(metadata_dict, f, indent=2, ensure_ascii=False)
        
        print(f"[SUCCESS] Metadata saved: {os.path.basename(metadata_file_path)}")
        return True
        
    except Exception as e:
        print(f"[ERROR] Failed to save metadata: {e}")
        return False

def load_metadata(metadata_file_path):
    """
    Load metadata from JSON file
    
    Args:
        metadata_file_path (str): Path to metadata file
        
    Returns:
        dict: Metadata dictionary or None if failed
    """
    try:
        with open(metadata_file_path, 'r', encoding='utf-8') as f:
            metadata = json.load(f)
        
        print(f"[SUCCESS] Metadata loaded: {os.path.basename(metadata_file_path)}")
        return metadata
        
    except Exception as e:
        print(f"[ERROR] Failed to load metadata: {e}")
        return None

def create_encryption_metadata(file_info, encrypted_filename, iv, tag, encrypted_aes_key, sender="Unknown", receiver="Unknown"):
    """
    Create comprehensive metadata for encrypted file
    
    Args:
        file_info (dict): Original file information
        encrypted_filename (str): Name of encrypted file
        iv (bytes): Initialization vector
        tag (bytes): Authentication tag
        encrypted_aes_key (bytes): RSA-encrypted AES key
        sender (str): Sender identifier
        receiver (str): Receiver identifier
        
    Returns:
        dict: Complete metadata dictionary
    """
    timestamp = generate_timestamp()
    
    metadata = {
        # Encryption details
        "encrypted_file": encrypted_filename,
        "iv": iv.hex(),
        "tag": tag.hex(), 
        "encrypted_aes_key": encrypted_aes_key.hex(),
        "aes_key_size": len(iv) * 8 if len(iv) <= 4 else 256,  # Estimate key size
        
        # Original file information
        "original_file": {
            "name": file_info["name"],
            "extension": file_info["extension"],
            "size_bytes": file_info["size_bytes"],
            "size_mb": file_info["size_mb"],
            "mime_type": file_info["mime_type"],
            "encoding": file_info["encoding"],
            "sha256_hash": file_info["sha256_hash"]
        },
        
        # Encryption session details
        "encryption_info": {
            "algorithm": "AES-GCM + RSA-OAEP",
            "aes_mode": "GCM",
            "rsa_padding": "OAEP",
            "timestamp": timestamp,
            "created_at": datetime.now().isoformat()
        },
        
        # Communication details
        "communication": {
            "sender": sender,
            "receiver": receiver
        },
        
        # Metadata version for future compatibility
        "metadata_version": "1.0"
    }
    
    return metadata

def list_files_in_folder(folder_path, extensions=None):
    """
    List files in a folder with optional extension filtering
    
    Args:
        folder_path (str): Path to folder
        extensions (list): List of file extensions to include (e.g., ['.txt', '.pdf'])
        
    Returns:
        list: List of file paths
    """
    try:
        folder_path = Path(folder_path)
        
        if not folder_path.exists():
            print(f"[WARNING] Folder does not exist: {folder_path}")
            return []
        
        files = []
        for file_path in folder_path.iterdir():
            if file_path.is_file():
                if extensions is None or file_path.suffix.lower() in extensions:
                    files.append(str(file_path))
        
        print(f"[INFO] Found {len(files)} file(s) in {folder_path.name}/")
        return sorted(files)
        
    except Exception as e:
        print(f"[ERROR] Failed to list files: {e}")
        return []

def get_folder_summary(folder_path):
    """
    Get summary of files in a folder
    
    Args:
        folder_path (str): Path to folder
        
    Returns:
        dict: Folder summary with file count and total size
    """
    try:
        folder_path = Path(folder_path)
        
        if not folder_path.exists():
            return {"file_count": 0, "total_size_mb": 0, "files": []}
        
        files = []
        total_size = 0
        
        for file_path in folder_path.iterdir():
            if file_path.is_file() and not file_path.name.startswith('.'):
                size = file_path.stat().st_size
                total_size += size
                files.append({
                    "name": file_path.name,
                    "size_bytes": size,
                    "size_mb": round(size / (1024 * 1024), 3)
                })
        
        return {
            "file_count": len(files),
            "total_size_mb": round(total_size / (1024 * 1024), 3),
            "files": sorted(files, key=lambda x: x["name"])
        }
        
    except Exception as e:
        print(f"[ERROR] Failed to get folder summary: {e}")
        return {"file_count": 0, "total_size_mb": 0, "files": []}

def print_folder_contents(folders):
    """
    Print summary of all project folders
    
    Args:
        folders (dict): Dictionary of folder paths
    """
    print("\n" + "="*60)
    print("[FOLDERS] PROJECT FOLDER CONTENTS")
    print("="*60)
    
    for folder_name, folder_path in folders.items():
        summary = get_folder_summary(folder_path)
        print(f"\n[{folder_name.upper()}] ({summary['file_count']} files, {summary['total_size_mb']} MB)")
        
        for file_info in summary['files'][:5]:  # Show first 5 files
            print(f"   • {file_info['name']} ({file_info['size_mb']} MB)")
        
        if len(summary['files']) > 5:
            print(f"   ... and {len(summary['files']) - 5} more files")

def cleanup_temp_files(folder_path, pattern="temp_*"):
    """
    Clean up temporary files matching a pattern
    
    Args:
        folder_path (str): Path to folder
        pattern (str): File pattern to match
        
    Returns:
        int: Number of files removed
    """
    try:
        import glob
        
        pattern_path = os.path.join(folder_path, pattern)
        temp_files = glob.glob(pattern_path)
        
        removed_count = 0
        for temp_file in temp_files:
            try:
                os.remove(temp_file)
                removed_count += 1
            except:
                pass
        
        if removed_count > 0:
            print(f"[INFO] Cleaned up {removed_count} temporary file(s)")
        
        return removed_count
        
    except Exception as e:
        print(f"[ERROR] Cleanup failed: {e}")
        return 0

if __name__ == "__main__":
    # Test file helper functionality
    print("🧪 Testing File Helper Utilities")
    print("=" * 40)
    
    # Test folder creation
    base_dir = "test_project"
    folders = create_project_folders(base_dir)
    
    # Test file operations
    test_file = os.path.join(folders['files'], "test.txt")
    with open(test_file, 'w') as f:
        f.write("Test content")
    
    # Test metadata
    test_metadata = {
        "test": True,
        "timestamp": generate_timestamp()
    }
    
    metadata_file = os.path.join(folders['metadata'], "test_metadata.json")
    save_metadata(test_metadata, metadata_file)
    
    loaded_metadata = load_metadata(metadata_file)
    
    if loaded_metadata and loaded_metadata.get("test"):
        print("[OK] File helper utilities test PASSED!")
    else:
        print("[ERROR] File helper utilities test FAILED!")
    
    # Print folder contents
    print_folder_contents(folders)
    
    # Cleanup
    import shutil
    if os.path.exists(base_dir):
        shutil.rmtree(base_dir)

File: backend/src/main/java/com/sentra/crypto/utils/FileHelper.java
package com.sentra.crypto.utils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import com.sentra.backend.runtime.Log;
import com.sentra.backend.runtime.JsonUtil;

public class FileHelper {
    public static Map<String, Path> createProjectFolders(Path baseDir) throws IOException {
        Map<String, Path> folders = new LinkedHashMap<>();
        folders.put("files", baseDir.resolve("files"));
        folders.put("encrypted", baseDir.resolve("encrypted"));
        folders.put("decrypted", baseDir.resolve("decrypted"));
        folders.put("metadata", baseDir.resolve("metadata"));
        folders.put("keys", baseDir.resolve("keys"));

        for (Map.Entry<String, Path> entry : folders.entrySet()) {
            Files.createDirectories(entry.getValue());
        }
        Log.info("BOOT", "folders", "initialized baseDir=" + baseDir);
        return folders;
    }

    public static String generateTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
    }

    public static String generateFilename(String originalName, String suffix, String extension, boolean includeTimestamp) {
        String baseName = originalName;
        int dotIndex = originalName.lastIndexOf('.');
        if (dotIndex > 0) {
            baseName = originalName.substring(0, dotIndex);
        }

        List<String> parts = new ArrayList<>();
        parts.add(baseName);
        if (suffix != null && !suffix.isEmpty()) {
            parts.add(suffix);
        }
        if (includeTimestamp) {
            parts.add(generateTimestamp());
        }

        return String.join("_", parts) + extension;
    }

    public static boolean saveMetadata(Map<String, Object> metadata, Path metadataFilePath) {
        try {
            Files.createDirectories(metadataFilePath.getParent());
            String json = JsonUtil.stringifyPretty(metadata);
            Files.writeString(metadataFilePath, json, StandardCharsets.UTF_8);
            Log.debug("META", "save", "file=" + metadataFilePath.getFileName());
            return true;
        } catch (Exception e) {
            Log.error("META", "save", "failed file=" + metadataFilePath.getFileName());
            return false;
        }
    }

    public static Map<String, Object> loadMetadata(Path metadataFilePath) {
        try {
            String content = Files.readString(metadataFilePath, StandardCharsets.UTF_8);
            Map<String, Object> metadata = JsonUtil.parseObject(content);
            Log.debug("META", "load", "file=" + metadataFilePath.getFileName());
            return metadata;
        } catch (Exception e) {
            Log.error("META", "load", "failed file=" + metadataFilePath.getFileName());
            return null;
        }
    }

    public static Map<String, Object> createEncryptionMetadata(
        Map<String, Object> fileInfo,
        String encryptedFilename,
        byte[] iv,
        byte[] tag,
        byte[] encryptedAesKey,
        String sender,
        String receiver
    ) {
        String timestamp = generateTimestamp();

        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("encrypted_file", encryptedFilename);
        metadata.put("iv", bytesToHex(iv));
        metadata.put("tag", bytesToHex(tag));
        metadata.put("encrypted_aes_key", bytesToHex(encryptedAesKey));
        metadata.put("aes_key_size", iv.length <= 4 ? iv.length * 8 : 256);

        Map<String, Object> originalFile = new LinkedHashMap<>();
        originalFile.put("name", fileInfo.get("name"));
        originalFile.put("extension", fileInfo.get("extension"));
        originalFile.put("size_bytes", fileInfo.get("size_bytes"));
        originalFile.put("size_mb", fileInfo.get("size_mb"));
        originalFile.put("mime_type", fileInfo.get("mime_type"));
        originalFile.put("encoding", fileInfo.get("encoding"));
        originalFile.put("sha256_hash", fileInfo.get("sha256_hash"));
        metadata.put("original_file", originalFile);

        Map<String, Object> encryptionInfo = new LinkedHashMap<>();
        encryptionInfo.put("algorithm", "AES-GCM + RSA-OAEP");
        encryptionInfo.put("aes_mode", "GCM");
        encryptionInfo.put("rsa_padding", "OAEP");
        encryptionInfo.put("timestamp", timestamp);
        encryptionInfo.put("created_at", LocalDateTime.now().toString());
        metadata.put("encryption_info", encryptionInfo);

        Map<String, Object> communication = new LinkedHashMap<>();
        communication.put("sender", sender);
        communication.put("receiver", receiver);
        metadata.put("communication", communication);

        metadata.put("metadata_version", "1.0");

        return metadata;
    }

    public static List<String> listFilesInFolder(Path folderPath) {
        try {
            if (!Files.exists(folderPath)) {
                Log.warn("FILES", "list", "missing folder=" + folderPath.getFileName());
                return Collections.emptyList();
            }
            List<String> files = new ArrayList<>();
            Files.list(folderPath).forEach(path -> {
                if (Files.isRegularFile(path)) {
                    files.add(path.toString());
                }
            });
            Collections.sort(files);
            Log.debug("FILES", "list", "folder=" + folderPath.getFileName() + " count=" + files.size());
            return files;
        } catch (Exception e) {
            Log.error("FILES", "list", "failed folder=" + folderPath.getFileName());
            return Collections.emptyList();
        }
    }

    public static Map<String, Object> getFolderSummary(Path folderPath) {
        try {
            if (!Files.exists(folderPath)) {
                return mapOf("file_count", 0, "total_size_mb", 0, "files", new ArrayList<>());
            }
            List<Map<String, Object>> files = new ArrayList<>();
            long totalSize = 0;
            Files.list(folderPath).forEach(path -> {
                try {
                    if (Files.isRegularFile(path) && !path.getFileName().toString().startsWith(".")) {
                        long size = Files.size(path);
                        Map<String, Object> info = new LinkedHashMap<>();
                        info.put("name", path.getFileName().toString());
                        info.put("size_bytes", size);
                        info.put("size_mb", Math.round((size / (1024.0 * 1024.0)) * 1000.0) / 1000.0);
                        files.add(info);
                    }
                } catch (Exception ex) {
                    Log.warn("FILES", "summary", "skip_file=" + path.getFileName() + " error=" + ex.getMessage());
                }
            });
            for (Map<String, Object> f : files) {
                totalSize += ((Number) f.get("size_bytes")).longValue();
            }

            files.sort(Comparator.comparing(o -> String.valueOf(o.get("name"))));

            Map<String, Object> summary = new LinkedHashMap<>();
            summary.put("file_count", files.size());
            summary.put("total_size_mb", Math.round((totalSize / (1024.0 * 1024.0)) * 1000.0) / 1000.0);
            summary.put("files", files);
            return summary;
        } catch (Exception e) {
            return mapOf("file_count", 0, "total_size_mb", 0, "files", new ArrayList<>());
        }
    }

    public static void printFolderContents(Map<String, Path> folders) {
        Log.debug("FILES", "summary", "project folder contents");

        for (Map.Entry<String, Path> entry : folders.entrySet()) {
            Map<String, Object> summary = getFolderSummary(entry.getValue());
            Log.debug("FILES", "summary", entry.getKey() + " files=" + summary.get("file_count") + " size_mb=" + summary.get("total_size_mb"));

            List<Map<String, Object>> files = (List<Map<String, Object>>) summary.get("files");
            int limit = Math.min(5, files.size());
            for (int i = 0; i < limit; i++) {
                Map<String, Object> f = files.get(i);
                Log.debug("FILES", "summary", "file=" + f.get("name"));
            }

            if (files.size() > 5) {
                Log.debug("FILES", "summary", "more_files=" + (files.size() - 5));
            }
        }
    }

    public static int cleanupTempFiles(Path folderPath, String pattern) {
        try {
            if (!Files.exists(folderPath)) {
                return 0;
            }
            int removed = 0;
            String prefix = pattern.replace("*", "");
            for (Path path : (Iterable<Path>) Files.list(folderPath)::iterator) {
                if (Files.isRegularFile(path) && path.getFileName().toString().startsWith(prefix)) {
                    try {
                        Files.delete(path);
                        removed++;
                    } catch (Exception ex) {
                        Log.warn("FILES", "cleanup", "skip_file=" + path.getFileName() + " error=" + ex.getMessage());
                    }
                }
            }
            if (removed > 0) {
                Log.debug("FILES", "cleanup", "removed=" + removed);
            }
            return removed;
        } catch (Exception e) {
            Log.error("FILES", "cleanup", "failed");
            return 0;
        }
    }

    private static String bytesToHex(byte[] data) {
        StringBuilder sb = new StringBuilder(data.length * 2);
        for (byte b : data) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    private static Map<String, Object> mapOf(Object... kv) {
        Map<String, Object> map = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) {
            map.put(String.valueOf(kv[i]), kv[i + 1]);
        }
        return map;
    }
}
